File: fe2-android/app/src/main/java/com/github/dedis/popstellar/repository/database/witnessing/WitnessingDao.java
package com.github.dedis.popstellar.repository.database.witnessing;

import androidx.room.*;

import com.github.dedis.popstellar.model.objects.WitnessMessage;

import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Single;

@Dao
public interface WitnessingDao {

  @Insert(onConflict = OnConflictStrategy.REPLACE)
  Completable insert(WitnessingEntity witnessingEntity);

  @Query("SELECT message FROM witness_messages WHERE lao_id = :laoId")
  Single<List<WitnessMessage>> getWitnessMessagesByLao(String laoId);
}

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/repository/database/witnessing/WitnessingDao.kt
package com.github.dedis.popstellar.repository.database.witnessing

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import com.github.dedis.popstellar.model.objects.WitnessMessage
import com.github.dedis.popstellar.model.objects.security.MessageID
import io.reactivex.Completable
import io.reactivex.Single

@Dao
interface WitnessingDao {
  @Insert(onConflict = OnConflictStrategy.REPLACE)
  fun insert(witnessingEntity: WitnessingEntity): Completable

  @Query("SELECT message FROM witness_messages WHERE lao_id = :laoId")
  fun getWitnessMessagesByLao(laoId: String): Single<List<WitnessMessage>?>

  @Query("DELETE FROM witness_messages WHERE lao_id = :laoId AND id IN (:filteredIds)")
  fun deleteMessagesByIds(laoId: String, filteredIds: Set<MessageID>): Completable
}
